### PR TITLE
Compute MoE load-balancing loss per layer to avoid giant one-hot materialization −99.7% @ 128k

### PR DESCRIPTION
**Pull request by qgallouedec**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48131)

`load_balancing_loss_func` concatenates the gate logits of **all layers**, then materializes `one_hot(topk(...), num_experts)` over the concatenation; an `[num_layers · tokens, top_k, num_experts]` int64 tensor (plus equally large expanded masks in the `attention_mask` path). Its peak memory grows with `num_layers × seq_len × top_k ×
num_experts` and dwarfs the model itself at long sequence lengths: training Qwen3-30B-A3B with
`output_router_logits=True` at 131K tokens/device tries to allocate a single 48 GiB tensor and
OOMs instantly.

internal context: https://huggingface.slack.com/archives/C0AEFSDG04X/p1787167930875279

Measured peak memory of just this function (Qwen3-30B-A3B shape — 48 layers, 128 experts, top-8, bf16 gate logits, H100):

Compute MoE load-balancing loss per layer to avoid giant one-hot materialization

| tokens/device | main | this PR |
|---|---|---|
| 8K | 5.3 GB | 0.02 GB |
| 32K | 20.5 GB | 0.06 GB |
| 64K | 40.7 GB | 0.12 GB |
| 128K | 81.2 GB | **0.25 GB** (−99.7%) |

image

### How

Accumulate per layer instead of concatenating: per-expert top-k assignment counts (`bincount` / mask-weighted `scatter_add_`) and routing-probability sums, normalized once at the end. No layer concatenation, no one-hot. The math is unchanged: summing the original `[top_k, num_experts]` token-fraction matrix over `top_k` (which the final `sum(tokens_per_expert * router_prob_per_expert)` does implicitly) equals the per-expert assignment count divided by the row count, and the row count is identical for every layer.

Models with independent hand-written variants (switch_transformers, nllb_moe) are out of scope.

### Verification

- Randomized equivalence vs the previous implementation, both `attention_mask=None` and masked, multiple shapes: **loss and `autograd.grad` w.r.t. gate logits match to 1e-6**:

```python
gates = tuple(torch.randn(B * S, E, requires_grad=True) for _ in range(L))
mask = (torch.rand(B, S) > 0.3).long()
for m in (None, mask):
    assert torch.allclose(old_fn(gates, E, K, m), load_balancing_loss_func(gates, E, K, m), atol=1e-6)
    # gradients likewise
```

- `tests/models/mixtral` and `tests/models/qwen3_moe` router tests pass; `utils/check_modular_conversion.py` clean.
- End-to-end: Qwen3-30B-A3B trains with TRL at 1M-token sequences (cp 8) where it previously OOMed on the 48 GiB allocation; `aux_loss` logs correctly and decreases.

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]
SGTM. Can you add the speed diff? But I think it might even be faster! 
TY that's a great catch + PT.

**Comment by qgallouedec** [comment]
Measured fwd+bwd of `load_balancing_loss_func` alone on synthetic router logits, 48 layers, 128 experts, top-8, batch 1, no attention mask, 1x H100 80GB, torch 2.13:

image


so tldr: a few ms slower below 8k (the per-layer loop is launch-bound), crossover at 16k, faster beyond (28.1 -> 20.1 ms at 32k, 55.8 -> 36.2 ms at 64k), and at 128k the old version OOMs the 80GB card by itself. The function runs once per forward, and even where it's slower we're talking ~7 ms extra against a fwd+bwd through a 48-layer MoE, which is orders of magnitude more, so nowhere near what dominates step time.

<details>
<summary>benchmark script</summary>

```python
import torch
import torch.nn.functional as F


def old_impl(gate_logits, num_experts, top_k, attention_mask=None):
    compute_device = gate_logits[0].device
    concatenated_gate_logits = torch.cat([g.to(compute_device) for g in gate_logits], dim=0)
    routing_weights = F.softmax(concatenated_gate_logits, dim=-1)
    _, selected_experts = torch.topk(routing_weights, top_k, dim=-1)
    expert_mask = F.one_hot(selected_experts, num_experts)
    tokens_per_expert = torch.mean(expert_mask.float(), dim=0)
    router_prob_per_expert = torch.mean(routing_weights, dim=0)
    overall_loss = torch.sum(tokens_per_expert * router_prob_per_expert.unsqueeze(0))
    return overall_loss * num_experts


def new_impl(gate_logits, num_experts, top_k, attention_mask=None):
    compute_device = gate_logits[0].device
    tokens_per_expert_sum = torch.zeros(num_experts, dtype=torch.float32, device=compute_device)
    router_prob_sum = torch.zeros(num_experts, dtype=torch.float32, device=compute_device)
    total_rows = 0.0
    for layer_gate in gate_logits:
        routing_weights = F.softmax(layer_gate.to(compute_device), dim=-1)
        _, selected_experts = torch.topk(routing_weights, top_k, dim=-1)
        tokens_per_expert_sum = (
            tokens_per_expert_sum + torch.bincount(selected_experts.reshape(-1), minlength=num_experts).float()
        )
        router_prob_sum = router_prob_sum + routing_weights.float().sum(dim=0)
        total_rows = total_rows + routing_weights.shape[0]
    tokens_per_expert = tokens_per_expert_sum / total_rows
    router_prob_per_expert = router_prob_sum / total_rows
    overall_loss = torch.sum(tokens_per_expert * router_prob_per_expert.unsqueeze(0))
    return overall_loss * num_experts


def bench(fn, gate_logits, num_experts, top_k, iters=20):
    start, end = torch.cuda.Event(enable_timing=True), torch.cuda.Event(enable_timing=True)
    for _ in range(3):  # warmup
        fn(gate_logits, num_experts, top_k).backward()
    torch.cuda.synchronize()
    start.record()
    for _ in range(iters):
        fn(gate_logits, num_experts, top_k).backward()
    end.record()
    torch.cuda.synchronize()
    return start.elapsed_time(end) / iters


layers, experts, top_k = 48, 128, 8
for seq in (1024, 2048, 4096, 8192, 16384, 32768, 65536, 131072):
    for label, fn in (("old", old_impl), ("new", new_impl)):
        torch.cuda.empty_cache()
        gate_logits = tuple(
            torch.randn(seq, experts, device="cuda", dtype=torch.float32, requires_grad=True) for _ in range(layers)
        )
        try:
            print(label, seq, f"{bench(fn, gate_logits, experts, top_k):.2f} ms", flush=True)
        except torch.OutOfMemoryError:
            print(label, seq, "OOM", flush=True)
        del gate_logits
```

</details>
